### PR TITLE
[blockindex] restrict height for sgdindexer during write and read operation

### PR DESCRIPTION
**Pull request by envestcc**
# Description
Motivation:
- Strictly limiting the blocks during PutBlock can avoid block skipping issues
-  Specifying the height during queries can help detect errors in the indexer at an earlier stage

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/execution/evm/evm.go [hunk 1]
```diff
@@ -303,8 +303,11 @@ func processSGD(ctx context.Context, sm protocol.StateManager, execution *action
 	if execution.Contract() == action.EmptyAddress {
 		return nil, 0, nil
 	}
-
-	receiver, percentage, ok, err := sgd.CheckContract(ctx, execution.Contract())
+	height, err := sm.Height()
+	if err != nil {
+		return nil, 0, err
+	}
+	receiver, percentage, ok, err := sgd.CheckContract(ctx, execution.Contract(), height)
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:310]: should be `height-1`? pls confirm

File: blockindex/sgd_indexer.go [hunk 2]
```diff
@@ -285,9 +285,12 @@ func (sgd *sgdRegistry) StartHeight() uint64 {
 
 // PutBlock puts a block into SGDIndexer
 func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
-	if blk.Height() < sgd.startHeight {
+	if ignore, err := sgd.validateBlock(blk); err != nil {
+		return err
+	} else if ignore {
```
Comment (dustinxie) [inline blockindex/sgd_indexer.go:290]: this is not very clean, i think just make it same as `contractstaking.Indexer.PutBlock()` which you fixed in #3924

File: blockindex/sgd_indexer.go [hunk 3]
```diff
@@ -285,9 +285,21 @@ func (sgd *sgdRegistry) StartHeight() uint64 {
 
 // PutBlock puts a block into SGDIndexer
 func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
-	if blk.Height() < sgd.startHeight {
+	tipHeight, err := sgd.Height()
```
Comment (Liuhaai) [inline blockindex/sgd_indexer.go:288]: wrap into a subfunc

File: blockindex/sgd_indexer.go [hunk 4]
```diff
@@ -481,6 +499,21 @@ func (sgd *sgdRegistry) FetchContracts(ctx context.Context) ([]*SGDIndex, error)
 	return sgdIndexes, nil
 }
 
+func (sgd *sgdRegistry) validateQueryHeight(height uint64) error {
+	// 0 means latest height
+	if height == 0 {
+		return nil
+	}
+	tipHeight, err := sgd.Height()
```
Comment (Liuhaai) [inline blockindex/sgd_indexer.go:507]: sgd.height()

File: blockindex/sgd_indexer.go (hunk 3 above)
Comment (envestcc) [reply blockindex/sgd_indexer.go:288]: wrap a function to calculate `expectHeight`

File: blockindex/sgd_indexer.go [hunk 5]
```diff
@@ -481,6 +499,21 @@ func (sgd *sgdRegistry) FetchContracts(ctx context.Context) ([]*SGDIndex, error)
 	return sgdIndexes, nil
 }
 
+func (sgd *sgdRegistry) validateQueryHeight(height uint64) error {
+	// 0 means latest height
+	if height == 0 {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply blockindex/sgd_indexer.go:504]: generally unlikely to occur, this is merely implemented as a feature to support querying the latest height data.

File: action/protocol/execution/evm/evm.go [hunk 6]
```diff
@@ -303,8 +303,11 @@ func processSGD(ctx context.Context, sm protocol.StateManager, execution *action
 	if execution.Contract() == action.EmptyAddress {
 		return nil, 0, nil
 	}
-
-	receiver, percentage, ok, err := sgd.CheckContract(ctx, execution.Contract())
+	height, err := sm.Height()
+	if err != nil {
+		return nil, 0, err
+	}
+	receiver, percentage, ok, err := sgd.CheckContract(ctx, execution.Contract(), height-1)
```
Comment (millken) [inline action/protocol/execution/evm/evm.go:310]: if height == 0, `height - 1` will overflow

File: blockindex/sgd_indexer.go [hunk 7]
```diff
@@ -285,14 +281,10 @@ func (sgd *sgdRegistry) StartHeight() uint64 {
 
 // PutBlock puts a block into SGDIndexer
 func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
-	tipHeight, err := sgd.Height()
+	expectHeight, err := sgd.expectHeight()
```
Comment (dustinxie) [inline blockindex/sgd_indexer.go:284]: nit: also do this for contract staking indexer? to keep it consistent

File: action/protocol/execution/evm/evm.go (hunk 6 above)
Comment (envestcc) [reply action/protocol/execution/evm/evm.go:310]: a good reminder, but it is a hard-fork after Quebec at least, so height=0 will not occur.

**Review by millken — APPROVED** [review]

File: action/protocol/execution/evm/evm.go (hunk 6 above)
Comment (Liuhaai) [reply action/protocol/execution/evm/evm.go:310]: > if height == 0, `height - 1` will overflow

validate the height >= 1 before calling CheckContract

File: blockindex/sgd_indexer.go (hunk 7 above)
Comment (Liuhaai) [reply blockindex/sgd_indexer.go:284]: sorry i mean wrapping the L284-294 into a subfunc

**Review by dustinxie — APPROVED** [review]

**Pull request merged**